src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+        } else if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index not valid");
+        } else {
+            if (size == elementData.length) {
+                grow();
+            }
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (elementData.length - size < list.size()) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
+        if (checkIndex(index)) {
+            return elementData[index];
+        }
         return null;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (checkIndex(index)) {
+            elementData[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = null;
+        if (checkIndex(index)) {
+            value = elementData[index];
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+            elementData[--size] = null;
+        }
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T value = null;
+        int index = -1;
+        if (null == element) {
+            for (int i = 0; i < size; i++) {
+                if (elementData[i] == null) {
+                    index = i;
+                    break;
+                }
+            }
+            value = checkRemoveIndex(index);
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (element.equals(elementData[i])) {
+                    index = i;
+                    break;
+                }
+            }
+            value = checkRemoveIndex(index);
```

Comment:
Too heavy construction with 2 cycles with brakes and repeatable logic, please rework using Objects.equals() method which will allow to perform null-safe check and use one cycle for both null and non-null cases